rust/kernel/lib.rs

```diff
@@ -69,6 +70,34 @@ pub trait Module: Sized + Sync {
     fn init(module: &'static ThisModule) -> error::Result<Self>;
 }
 
+/// A module that is pinned and initialised in-place.
+pub trait InPlaceModule: Sync + Send {
+    /// The type that implements the pinned initialisation of the module.
+    type Init: init::PinInit<Self, error::Error>;
+
+    /// Creates an initialiser for the module.
+    ///
+    /// It is called when the module is loaded.
+    fn init(module: &'static ThisModule) -> error::Result<Self::Init>;
+}
+
+impl<T: Module> InPlaceModule for T {
+    type Init = impl init::PinInit<Self, error::Error>;
+
+    fn init(module: &'static ThisModule) -> error::Result<Self::Init> {
+        let m = <Self as Module>::init(module)?;
+
+        let initer = move |slot: *mut Self| {
+            // SAFETY: `slot` is valid for write per the contract with `pin_init_from_closure`.
+            unsafe { slot.write(m) };
+            Ok(())
+        };
+
+        // SAFETY: On success, `initer` always fully initialises an instance of `Self`.
+        Ok(unsafe { init::pin_init_from_closure(initer) })
```

Comment:
Thanks to this I found out that a thing in pin-init is missing something [^1], with that fixed this can become
```rust
<Self as Module>::init(module)
```

[^1]: the `impl<T> Init<T> for T` should be `impl<T, E> Init<T, E> for T`.